services-api/src/main/java/io/scalecube/services/methods/ServiceMethodInvoker.java

```diff
@@ -106,6 +122,38 @@ public ServiceMethodInvoker(
     return throwable != null ? Mono.error(throwable) : result;
   }
 
+  private Object[] prepareArguments(Object request, Object principal) {
+    Object[] arguments = new Object[methodInfo.parameterCount()];
+    Object principalArg = principal.equals(NO_PRINCIPAL) ? null : principal;
+
+    if (methodInfo.requestType() != Void.TYPE) {
+      arguments[0] = request;
+    } else {
+      arguments[0] = principalArg;
+    }
+
+    if (methodInfo.parameterCount() > 1) {
+      arguments[1] = principalArg;
+    }
+    return arguments;
+  }
+
+  private Mono<Object> authenticate(ServiceMessage message) {
+    return Mono.defer(
+        () -> {
+          if (!methodInfo.isAuth()) {
+            return Mono.empty();
+          }
+          if (authenticator == null) {
+            throw new UnauthorizedException("Authenticator not found");
+          }
+          return authenticator
+              .authenticate(message)
+              .onErrorMap(th -> new UnauthorizedException(th.getMessage()));
```

Comment:
Not all exceptions would have message. Maybe better to pass `th` itself?

Reply:
Done https://github.com/scalecube/scalecube-services/pull/624/commits/a02e7be740b3671bff33430f668c35d6969bebbd